allure-robotframework/src/listener.py

```diff
@@ -0,0 +1,152 @@
+from allure_commons.logger import AllureFileLogger
+from allure_commons.types import AttachmentType
+from allure_commons import plugin_manager
+from robot.libraries.BuiltIn import BuiltIn
+from allure_robotframework.constants import RobotKeywordType, RobotLogLevel
+from allure_robotframework import utils
+import os
+
+
+# noinspection PyPep8Naming
+class allure_robotframework(object):
+    ROBOT_LISTENER_API_VERSION = 2
+    DEFAULT_OUTPUT_PATH = os.path.join('output', 'allure')
+    LOG_MESSAGE_FORMAT = '{full_message}<p><b>[{level}]</b> {message}</p>'
+    FAIL_MESSAGE_FORMAT = '{full_message}<p style="color: red"><b>[{level}]</b> {message}</p>'
+
+    def __init__(self, logger_path=DEFAULT_OUTPUT_PATH):
+        self.reporter = AllureReporter()
+        self.logger = AllureFileLogger(logger_path)
+        self.stack = []
+        self.items_log = {}
+        self.pool_id = None
+        plugin_manager.register(self.reporter)
+        plugin_manager.register(self.logger)
+
+    def start_suite(self, name, attributes):
+        if not self.pool_id:
+            self.pool_id = BuiltIn().get_variable_value('${PABOTEXECUTIONPOOLID}')
+            self.pool_id = int(self.pool_id) if self.pool_id else 0
+        self.start_new_group(name, attributes)
+
+    def end_suite(self, name, attributes):
+        self.stop_current_group()
+
+    def start_test(self, name, attributes):
+        self.start_new_group(name, attributes)
+        self.start_new_test(name, attributes)
+
+    def end_test(self, name, attributes):
+        self.stop_current_test(name, attributes)
+        self.stop_current_group()
+
+    def start_keyword(self, name, attributes):
+        self.start_new_keyword(name, attributes)
+
+    def end_keyword(self, name, attributes):
+        self.end_current_keyword(name, attributes)
+
+    def log_message(self, message):
+        level = message.get('level')
+        if level == RobotLogLevel.FAIL:
+            self.reporter.get_item(self.stack[-1]).statusDetails = StatusDetails(message=message.get('message'))
+        self.append_message_to_last_item_log(message, level)
+
+    # listener event ends
+    def start_new_group(self, name, attributes):
+        uuid = uuid4()
+        if self.stack:
+            parent_suite = self.reporter.get_item(self.stack[-1])
+            parent_suite.children.append(uuid)
+        self.stack.append(uuid)
+        suite = TestResultContainer(uuid=uuid,
+                                    name=name,
+                                    description=attributes.get('doc'),
+                                    start=now())
+        self.reporter.start_group(uuid, suite)
+
+    def stop_current_group(self):
+        uuid = self.stack.pop()
+        self.reporter.stop_group(uuid, stop=now())
+
+    def start_new_test(self, name, attributes):
+        uuid = uuid4()
+        self.reporter.get_item(self.stack[-1]).children.append(uuid)
+        self.stack.append(uuid)
+        test_case = TestResult(uuid=uuid,
```

Comment:
I think this is a right place to generate a historyId from attributes['longname']

Reply:
Done